mono/metadata/metadata.c

```diff
@@ -3195,7 +3160,29 @@ select_container (MonoGenericContainer *gc, MonoTypeEnum type)
 	return gc;
 }
 
-/* 
+// FIXME: Is the distinction between method_container and class_container necessary? It's possible method_containers are never created.
+MonoGenericContainer *
+get_anonymous_container_for_image (MonoImage *image, gboolean is_mvar)
+{
+	MonoGenericContainer **container_pointer;
+	if (is_mvar)
+		container_pointer = &image->anonymous_generic_method_container;
+	else
+		container_pointer = &image->anonymous_generic_class_container;
+	// This container has never been created; make it now.
+	if (!*container_pointer)
+	{
+		// Note this is never deallocated anywhere-- it exists for the lifetime of the image it's allocated from
+		*container_pointer = mono_image_alloc0 (image, sizeof (MonoGenericContainer));
```

Comment:
This is publishing uninitialized data as there's no locking around the parsing code.
It might cause other threads to see partially initialized data.

The right way is to init on a temporary and only at the end store it to the right place.

Then the question is whether the anon container must be unique or not.

If it doesn't need to be unique, a store_release is enough (and it might cause duplicates).

If it does, CAS the value in.